### PR TITLE
Pre mainnet oracle fixes

### LINKED ISSUE #248: pre-mainnet review: LidoOracle
- [x] **error** https://github.com/lidofinance/lido-dao/blob/ad4b2f634850e46375bf5c8717abf0e49cdce3be/contracts/0.4.24/oracle/LidoOracle.sol#L155-L161

`_member`'s data might be in epochs other than `lastReportedEpochId`, namely, in any epoch in `getCurrentReportableEpochs()`. This data must be discarded as well. The simplest solution is to discard current `gatheredEpochData` altogether. Alternatively, `MIN_REPORTABLE_EPOCH_ID_VALUE_POSITION` may be set to `getCurrentEpochId()` and member data removal may be performed for that epoch.

- [x] **error** https://github.com/lidofinance/lido-dao/blob/ad4b2f634850e46375bf5c8717abf0e49cdce3be/contracts/0.4.24/oracle/LidoOracle.sol#L187

Why is the comparison strict rather than non-strict? The comparison can be omitted altogether since an epoch without reports has a zero as `reportsBitMask`. That makes finalization impossible.

- [x] **error** https://github.com/lidofinance/lido-dao/blob/ad4b2f634850e46375bf5c8717abf0e49cdce3be/contracts/0.4.24/oracle/LidoOracle.sol#L188

The assignment is futile since the local variable `minReportableEpochId` is not used afterwards.

- [x] warning https://github.com/lidofinance/lido-dao/blob/ad4b2f634850e46375bf5c8717abf0e49cdce3be/contracts/0.4.24/oracle/LidoOracle.sol#L214

Because of the reasons mentioned above it makes no sense to track this value.

- [x] comment https://github.com/lidofinance/lido-dao/blob/ad4b2f634850e46375bf5c8717abf0e49cdce3be/contracts/0.4.24/oracle/LidoOracle.sol#L101-L104

Copy-pasted from `setBeaconSpec`. It makes sense to move the checks to `_setBeaconSpec`.

### PR DESCRIPTION
**Pull request by lxzrv**
Closes #248

### CODE REVIEW

**Review by ongrid — CHANGES_REQUESTED** [review]
@lxzrv need rebase on develop's HEAD and resolve conflicts.

**Review by ongrid — APPROVED** [review]

**Review by Eenae — APPROVED** [review]

File: contracts/0.4.24/oracle/LidoOracle.sol [hunk 1]
```diff
@@ -144,20 +139,18 @@ contract LidoOracle is ILidoOracle, AragonApp {
         uint256 index = _getMemberId(_member);
         require(index != MEMBER_NOT_FOUND, "MEMBER_NOT_FOUND");
 
-        uint256 lastReportedEpochId = (
-            LAST_REPORTED_EPOCH_ID_POSITION.getStorageUint256()
-        );
+        uint256 maxReportedEpochId = MAX_REPORTED_EPOCH_ID_POSITION.getStorageUint256();
 
-        MIN_REPORTABLE_EPOCH_ID_POSITION.setStorageUint256(lastReportedEpochId);
+        MIN_REPORTABLE_EPOCH_ID_POSITION.setStorageUint256(maxReportedEpochId);
```
Comment (dechjo) [inline contracts/0.4.24/oracle/LidoOracle.sol:144]: mistype?.. investigating...

File: contracts/0.4.24/oracle/LidoOracle.sol [hunk 2]
```diff
@@ -174,18 +167,16 @@ contract LidoOracle is ILidoOracle, AragonApp {
...
+        uint256 maxReportedEpochId = MAX_REPORTED_EPOCH_ID_POSITION.getStorageUint256();
 
-        assert(lastReportedEpochId <= getCurrentEpochId());
+        assert(maxReportedEpochId <= getCurrentEpochId());
 
-        if (lastReportedEpochId > minReportableEpochId) {
-            minReportableEpochId = lastReportedEpochId;
-            _tryPush(lastReportedEpochId);
+        if (maxReportedEpochId >= minReportableEpochId) {
+            if (maxReportedEpochId != minReportableEpochId) {
```
[the remark this answers is not in the document]
Comment (vshvsh) [reply contracts/0.4.24/oracle/LidoOracle.sol:176]: You're right it's not a good idea. Up to 7200 iterations in a good day, unlimited in a bad network weather. 

Ditching old data on quorum change is acceptable, though. Needs to be documented but acceptable.

[the remark this answers is not in the document]
Comment (dechjo) [reply contracts/0.4.24/oracle/LidoOracle.sol:176]: @vshvsh, where do you think to document it? https://github.com/lidofinance/lido-dao/blob/develop/docs/protocol-levers.md#lidooraclesol?

**Comment by dechjo** [comment]
Manually merged to `oracle-v2` branch, see #292.

**Pull request closed without merging**